packages/horizon/ignition/modules/core/HorizonStaking.ts

```diff
@@ -0,0 +1,34 @@
+import { buildModule } from '@nomicfoundation/hardhat-ignition/modules'
+
+import GraphPeripheryModule from '../periphery'
+import HorizonProxiesModule from './HorizonProxies'
+import HorizonStakingExtensionModule from './HorizonStakingExtension'
+
+export default buildModule('HorizonStaking', (m) => {
+  const { Controller, GraphProxyAdmin, PeripheryRegistered } = m.useModule(GraphPeripheryModule)
+  const { HorizonStakingProxy, HorizonRegistered } = m.useModule(HorizonProxiesModule)
+  const { HorizonStakingExtension } = m.useModule(HorizonStakingExtensionModule)
+
+  const subgraphServiceAddress = m.getParameter('subgraphServiceAddress')
+
+  // Deploy HorizonStaking implementation
+  const HorizonStakingImplementation = m.contract('HorizonStaking',
+    [
+      Controller,
+      HorizonStakingExtension,
+      subgraphServiceAddress,
+    ],
+    {
+      after: [PeripheryRegistered, HorizonRegistered],
+    },
+  )
+
+  // Upgrade proxy to implementation contract
+  const upgradeCall = m.call(GraphProxyAdmin, 'upgrade', [HorizonStakingProxy, HorizonStakingImplementation])
+  const acceptCall = m.call(GraphProxyAdmin, 'acceptProxy', [HorizonStakingImplementation, HorizonStakingProxy], { after: [upgradeCall] })
+
+  // Load contract with implementation ABI
+  const HorizonStaking = m.contractAt('HorizonStaking', HorizonStakingProxy, { after: [acceptCall], id: 'HorizonStaking_Instance' })
```

Comment:
Is `contractAt` forcing you to use explicit `id`s too often? Or are you setting it here for a stylistic preference?

Reply:
Oh, I guess this happens when you do "deploy implementation", "contractAt implementation". I'll think about how to improve this. 

Maybe with something like what you did here: https://github.com/graphprotocol/contracts/pull/1025#discussion_r1761950531

Reply:
yeah had to set it because of that! 

Perhaps Ignition could detect this pattern:
```
const proxy = m.contract('Proxy')
const implementation = m.contract('ContractName')
const proxyWithImplementationABI = m.contractAt('ContractName', proxy)
```

And automatically add something to avoid id collision on the contractAt. But it's also not a big deal I think.

Reply:
Thanks for confirming, I created an issue to keep track of this: https://github.com/NomicFoundation/hardhat-ignition/issues/814